docs/architecture/grpc-for-wcf-developers/service-mesh.md

```diff
@@ -1,11 +1,123 @@
 ---
 title: Service meshes - gRPC for WCF Developers
-description: TO BE WRITTEN
+description: Using a service mesh to route and balance requests to gRPC services in a Kubernetes cluster.
 author: markrendle
 ms.date: 09/02/2019
 ---
 
 # Service meshes
 
+A service mesh is an infrastructure component that takes control of routing service requests within a network. Service meshes can handle all kinds of network-level concerns within a Kubernetes cluster, including:
+
+- Service discovery
+- Load balancing
+- Fault tolerance
+- Encryption
+- Monitoring
+
+Kubernetes service meshes work by adding an extra container, called a *sidecar proxy*, to each pod included in the mesh. The proxy takes over handling all inbound and outbound network requests, allowing configuration and management of networking matters to be kept separate from the application containers and, in many cases, without requiring any changes to the application code.
```

Comment:
As this is a book about grpc for WCF developers and not just a book about grpc, would it be worth mentioning this is similar to WCF routing services so if you currently use that, this would be your replacement when migrating to gRPC?

Reply:
I don't think service meshes do anything like the stuff that WCF Routing could do, with the possible exception of rerouting failed requests to alternative endpoints.

Reply:
From the routing services [introduction documentation](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/routing-introduction):

> The Routing Service provides a generic pluggable SOAP intermediary that is capable of routing messages based on message content. With the Routing Service, you can create complex routing logic that allows you to implement scenarios such as service aggregation, service versioning, priority routing, and multicast routing. The Routing Service also provides error handling that allows you to set up lists of backup endpoints, to which messages are sent if a failure occurs when sending to the primary destination endpoint.

So it handles Load balancing, fault tolerance, although not listed here, it can handle the encryption for you as forward binding doesn't need to match frontend binding. From what I've read about service mesh, routing services are a lot more powerful but for many use cases, the migration path does seem like a service mesh would be the answer. Am I missing something?